ambari-server/src/main/java/org/apache/ambari/server/topology/addservice/AddServiceInfo.java

```diff
@@ -65,11 +75,11 @@ public AddServiceInfo(
 
   public AddServiceInfo withLayoutRecommendation(Map<String, Map<String, Set<String>>> services,
                                                  LayoutRecommendationInfo recommendation) {
-    return new AddServiceInfo(request, clusterName, stack, config, kerberosDescriptor, stages, services, recommendation);
+    return new AddServiceInfo(request, clusterName, stages, stack, config, services, kerberosDescriptor, recommendation);
   }
 
   public AddServiceInfo withConfig(Configuration newConfig) {
-    return new AddServiceInfo(request, clusterName, stack, newConfig, kerberosDescriptor, stages, newServices, recommendationInfo);
+    return new AddServiceInfo(request, clusterName, stages, stack, newConfig, newServices, kerberosDescriptor, recommendationInfo);
   }
 
   @Override
```

Comment:
I thought parameters that are "more required" should appear earlier to avoid `(..., null, null, someValue, ...)` as much as possible.  Of course a builder would be even better with this many parameters...

Reply:
Are all the parameters required? If not, then I'd definitely vote for a builder here.

Reply:
Created follow-up issue:
https://issues.apache.org/jira/browse/AMBARI-25061